File: chanclol/bot.py
import asyncio
import discord
import logging
import math

from riotapi.riotapi import RiotApi
import parser
import message_formatter as mft
from database.database_bot import DatabaseBot
from timed_executor import TimedExecutor
from guild import Guild
from player import Player
from riot_id import RiotId

logger = logging.getLogger(__name__)


# The bot: receives commands from the discord client
# and processes them. It runs an infinite loop that checks the
# in-game status of all the players registered for each guild
class Bot:

    def __init__(
        self,
        client,
        database_filename_bot: str,
        offline_threshold_mins: int,
        timeout_online_secs: int,
        timeout_offline_secs: int,
        main_loop_cycle_secs: int,
        riotapi_housekeeping_cycle_hrs: int,
        riotapi_key: str,
        database_filename_riotapi: str,
        restrictions: list[dict],
    ):

        # Keep a copy of the client
        self.client = client
        # Create the database
        self.database = DatabaseBot(database_filename_bot)
        # Initialise guilds and players from the database, if any
        self.guilds = self.database.get_guilds()
        self.players = self.database.get_players()
        # The bot executes a function by which it tells the riot API
        # which of the player ids to keep, so that the others can be purged
        self.riotapi_housekeeping_executor = TimedExecutor(
            riotapi_housekeeping_cycle_hrs * 60 * 60 * 1000, self.riotapi_housekeeping
        )
        # Timeouts for online and offline players
        self.offline_threshold_mins = offline_threshold_mins
        self.timeout_online_secs = timeout_online_secs
        self.timeout_offline_secs = timeout_offline_secs
        # Set default offline timeouts for all players
        for player in self.players.values():
            player.stopwatch.set_timeout(self.timeout_offline_secs * 1000)
        # Main loop cycle
        self.main_loop_cycle_secs = main_loop_cycle_secs
        # Riot API
        self.riot_api = RiotApi(riotapi_key, database_filename_riotapi, restrictions)

    # Main entry point for all messages
    async def receive(self, message):

        # Reject my own messages
        if message.author == self.client.user:
            logger.debug("Rejecting message sent by myself")
            return

        # For the time being, ignore messages sent on private channels
        if not message.guild:
            logger.info("Ignoring private message")
            await message.channel.send(
                "For the time being, I am ignoring private messages"
            )
            return

        # Register a guild in case it does not yet exist
        if not message.guild.id in self.guilds:
            logger.info(f"Initialising guild {message.guild.id}")
            self.guilds[message.guild.id] = Guild(message.guild.id, message.channel.id)
            self.database.add_guild(message.guild.id, message.channel.id)
            logger.info("Sending welcome message")
            response = mft.welcome(message.channel.name)
            await message.channel.send(content=response.content, embed=response.embed)
        guild = self.guilds[message.guild.id]

        # Parse the input provided and call the appropriate function
        logger.debug(f"Message received: {message.content}")
        parsed_input = parser.Parser(message.content)
        match parsed_input.code:
            case parser.Result.NO_BOT_PREFIX:
                logger.debug("Rejecting message not intended for the bot")
            case parser.Result.OK:
                logger.info(f"Command understood: {message.content}")
                match parsed_input.command:
                    case parser.Command.REGISTER:
                        responses = await self.register(
                            RiotId(
                                parsed_input.arguments[0], parsed_input.arguments[1]
                            ),
                            guild,
                        )
                    case parser.Command.UNREGISTER:
                        responses = await self.unregister(
                            RiotId(
                                parsed_input.arguments[0], parsed_input.arguments[1]
                            ),
                            guild,
                        )
                    case parser.Command.PRINT:
                        responses = await self.print(guild)
                    case parser.Command.CHANNEL:
                        responses = await self.channel(parsed_input.arguments[0], guild)
                    case parser.Command.HELP:
                        responses = [mft.create_help_message()]
                    case parser.Command.RANK:
                        responses = await self.rank(
                            RiotId(parsed_input.arguments[0], parsed_input.arguments[1])
                        )
                    case _:
                        raise ValueError("Command is not one of the possible ones")
                for response in responses:
                    await message.channel.send(
                        content=response.content, embed=response.embed
                    )
            case _:
                error_message = parsed_input.error_message
                if not error_message:
                    raise ValueError(
                        "Expected an error message in a parse failure case"
                    )
                else:
                    logger.info(
                        f"Wrong input: '{message.content}'. Reason: {error_message}"
                    )
                    response = mft.input_not_valid(error_message)
                    await message.channel.send(
                        content=response.content, embed=response.embed
                    )

    # Register the provided player in the provided guild, if possible
    async def register(self, riot_id: RiotId, guild: Guild) -> list[mft.Message]:

        # Get the puuid
        puuid = await self.riot_api.get_puuid(riot_id)
        if not puuid:
            logger.info(f"Riot has not provided a puuid for player {riot_id}")
            return [mft.no_response_riot_api(riot_id)]

        # Check if the player already belongs to the guild
        if puuid in guild.last_informed_game_ids:
            logger.info(f"Player {riot_id} is already registered")
            return [mft.player_already_registered(riot_id)]

        # Get rank info for this player
        league = await self.riot_api.get_leagues(puuid)
        if not league:
            logger.info(f"Could not get rank from Riot API for player {riot_id}")

        # Now it's safe to register the player
        logger.info(f"Registering player {riot_id}")
        # Add it to the list of players if necessary
        if not puuid in self.players:
            self.players[puuid] = Player(puuid)
            self.players[puuid].stopwatch.set_timeout(self.timeout_offline_secs * 1000)
        # Add to the guild
        guild.last_informed_game_ids[puuid] = None
        # Add to the database
        self.database.add_player_to_guild(puuid, guild.id)

        # Send a final message
        logger.info(f"Player {riot_id} has been registered")
        return [
            mft.player_registered(riot_id),
            mft.player_rank(riot_id, league),
        ]

    # Unregister a player from the guild
    async def unregister(self, riot_id: RiotId, guild: Guild) -> list[mft.Message]:

        # Get the puuid
        puuid = await self.riot_api.get_puuid(riot_id)
        if puuid == None:
            logger.info(f"Riot has not provided an id for player {riot_id}")
            return [mft.no_response_riot_api(riot_id)]

        # Check if the player was registered in the guild
        if puuid in guild.last_informed_game_ids:
            logger.info(f"Unregistering player {riot_id} from guild {guild.id}")
            del guild.last_informed_game_ids[puuid]
            # Remove from the database
            self.database.remove_player_from_guild(puuid, guild.id)
            logger.info(f"Player {riot_id} has been unregistered")
            # See if the player is in any other guild. If not, remove it completely
            guild_ids = self.get_guild_ids(puuid)
            if len(guild_ids) == 0:
                logger.info(f"Player {riot_id} will be completely removed")
                del self.players[puuid]
            return [mft.player_unregistered_correctly(riot_id)]
        else:
            logger.info(f"Player {riot_id} was not previously registered in this guild")
            return [mft.player_not_previously_registered(riot_id)]

    # Print the rank of a player, if it exists
    async def rank(self, riot_id: RiotId) -> list[mft.Message]:

        # Get the puuid
        puuid = await self.riot_api.get_puuid(riot_id)
        if not puuid:
            logger.info(f"Riot has not provided an id for player {riot_id}")
            return [mft.no_response_riot_api(riot_id)]

        # Get rank info for this player
        league = await self.riot_api.get_leagues(puuid)
        if not league:
            logger.info(f"Could not get rank info from Riot API for player {riot_id}")
            return [mft.no_response_riot_api(riot_id)]

        # Send a final message
        logger.info(f"Sending rank of player {riot_id}")
        return [mft.player_rank(riot_id, league)]

    # Change the channel where the in-game messages will be sent to
    async def channel(self, new_channel_name: str, guild: Guild) -> list[mft.Message]:

        discord_guild = self.client.get_guild(guild.id)
        # First make sure the channel does exist in the guild
        channel = discord.utils.get(discord_guild.channels, name=new_channel_name)
        if channel == None:
            logger.info(
                f"Cannot change channel to {new_channel_name} as it does not exist in the server"
            )
            return [mft.channel_does_not_exist(new_channel_name)]
        else:
            logger.info(f"Changing channel to use by this server to {new_channel_name}")
            guild.channel_id = channel.id
            self.database.set_channel_id(guild.id, guild.channel_id)
            logger.info(f"Channel changed to {new_channel_name}")
            return [mft.channel_changed(new_channel_name)]

    # Print the players currently registered in the provided guild
    async def print(self, guild: Guild) -> list[mft.Message]:

        # Very bad error if I cannot find the name of the channel to which
        # I am sending the messages
        channel_name = self.client.get_channel(guild.channel_id).name
        if channel_name == None:
            raise ValueError(
                f"Channel {channel_name} has not been found for this guild"
            )
        # Create list of player names in this guild
        riot_ids = []
        for puuid in guild.last_informed_game_ids:
            riot_id = await self.riot_api.get_riot_id(puuid)
            if riot_id:
                riot_ids.append(riot_id)
        return [mft.print_config(riot_ids, channel_name)]

    # Get a list of all the guild ids where the provided player is registered
    def get_guild_ids(self, puuid: str) -> list[int]:

        return [
            guild.id
            for guild in self.guilds.values()
            if puuid in guild.last_informed_game_ids
        ]

    # Main loop of the application
    async def loop(self):

        while True:
            # TODO: investigate if this wait time needs to be updated according to the number of users registered
            await asyncio.sleep(self.main_loop_cycle_secs)
            # Perform housekeeping of the riot API if necessary
            await self.riotapi_housekeeping_executor.execute()
            # Find a player to check in this iteration
            puuid = await self.select_player_to_check()
            if not puuid:
                continue
            riot_id = await self.riot_api.get_riot_id(puuid)
            if not riot_id:
                logger.error(f"Could not find a riot id for puuid {puuid}")
                continue
            # We have a player to check
            logger.debug(f"Checking in game status of player {riot_id}")
            # Make a request to Riot to check if the player is in game
            spectator = await self.riot_api.get_spectator(puuid)
            # At this point the request has been made in some form, so reset the stopwatch
            self.players[puuid].stopwatch.start()
            if not spectator:
                logger.debug(f"Spectator data for player {riot_id} is not available")
                # Player is offline, update the timeout if needed
                if self.players[puuid].update_check_timeout(
                    False,
                    self.offline_threshold_mins,
                    self.timeout_online_secs,
                    self.timeout_offline_secs,
                ):
                    logger.info(f"Player {riot_id} is now offline")
                continue
            # Player is online, update the timeout if needed
            if self.players[puuid].update_check_timeout(
                True,
                self.offline_threshold_mins,
                self.timeout_online_secs,
                self.timeout_offline_secs,
            ):
                logger.info(f"Player {riot_id} is now online")
            # Get the guilds where this player is registered
            guild_ids_player = self.get_guild_ids(puuid)
            for guild_id in guild_ids_player:
                guild = self.guilds[guild_id]
                # Get the last game that was informed for this player in this guild
                last_informed_game_id = guild.last_informed_game_ids[puuid]
                if last_informed_game_id == spectator.game_id:
                    logger.debug(
                        f"Message for player {riot_id} in guild {guild_id} for this game was already sent"
                    )
                else:
                    logger.info(
                        f"Player {riot_id} is in game and a message has to be sent to guild {guild_id}"
                    )
                    # Create the complete response
                    message = mft.in_game_message(puuid, riot_id, spectator)
                    if not message:
                        logger.error(
                            f"Could not produce an in game message for player with riot id {riot_id}"
                        )
                        continue
                    await self.send_message(message, guild.channel_id)
                    # Update the last informed game_id
                    logger.info("Updating last informed game id")
                    guild.last_informed_game_ids[puuid] = spectator.game_id
                    self.database.set_last_informed_game_id(
                        puuid, guild.id, spectator.game_id
                    )

    # Send the provided message in the provided channel id, if it exists
    async def send_message(self, message, channel_id: int) -> None:

        # Get the channel from channel id
        channel = self.client.get_channel(channel_id)
        if channel == None:
            logger.error(f"Could not find channel {channel_id} to send message to")
        else:
            await channel.send(content=message.content, embed=message.embed)

    # Perform housekeeping tasks on the Riot API
    async def riotapi_housekeeping(self) -> None:

        logger.info("Riot API housekeeping")
        # Create the list of player ids we want to keep
        player_ids_to_keep = list(self.players.keys())
        # Perform housekeeping in the Riot API
        await self.riot_api.housekeeping(player_ids_to_keep)

    # Select the player for which we will request the spectator data
    # As a rule of thumb, it will be the player for which we have not had this data
    # for the longer time.
    # Online players will be checked more frequently
    async def select_player_to_check(self) -> str | None:

        puuid_to_check: str | None = None
        best_time_behind = -math.inf
        for player in self.players.values():
            time_behind = player.stopwatch.time_behind()
            riot_id = await self.riot_api.get_riot_id(player.id)
            if time_behind > 0:
                if time_behind > best_time_behind:
                    # This player is available to be checked, and is more behind than the best,
                    # so it is the new best
                    logger.debug(f"Player {riot_id} is more behind")
                    best_time_behind = time_behind
                    puuid_to_check = player.id
                else:
                    # This player is available to check, but is less behind than the current best
                    logger.debug(f"Rejecting player {riot_id} as it is less behind")
            else:
                logger.debug(f"Timeout still not reached for player {riot_id}")
        # Here we in principle have selected the best option to check
        if not puuid_to_check:
            logger.debug("Not checking any player during this iteration")
            return None
        else:
            riot_id_to_check = await self.riot_api.get_riot_id(puuid_to_check)
            if not riot_id_to_check:
                logger.warning(
                    f"Selected player with id {puuid_to_check} to check, but could not find their riot id"
                )
            else:
                logger.debug(f"Will check status of player {riot_id_to_check}")
            return puuid_to_check

File: internal/bot/bot.go
package bot

import (
	"chanclol/internal/common"
	"chanclol/internal/riotapi"
	"errors"
	"fmt"
	"os"
	"os/signal"
	"sync"
	"syscall"
	"time"

	"github.com/bwmarrin/discordgo"
	"github.com/rs/zerolog/log"
)

type Guild struct {
	id                  string                           // id of the guild as handled by the discord library
	channelId           string                           // channel where in game messages will be sent
	lastInformedGameIds map[riotapi.Puuid]riotapi.GameId // map with last game ids informed for each player in the guild
}

type Player struct {
	id         riotapi.Puuid    // puuid of the player as handled by riot
	StopWatch  common.Stopwatch // a stopwatch that decides when to check the in game status
	lastOnline time.Time        // last time this player was seen online
}

// Check if the timeout has been reached for this player,
// depending on the player being online or offline.
// Return true if the status has changed (from online to offline or viceversa)
// and false otherwise
func (player *Player) UpdateCheckTimeout(online bool, offlineThreshold time.Duration, onlineTimeout time.Duration, offlineTimeout time.Duration) bool {

	currentTime := time.Now()
	if online {
		player.lastOnline = currentTime
		if player.StopWatch.Timeout != onlineTimeout {
			player.StopWatch.Timeout = onlineTimeout
			return true
		}
	} else {
		if currentTime.Sub(player.lastOnline) > offlineThreshold {
			if player.StopWatch.Timeout != offlineTimeout {
				player.StopWatch.Timeout = offlineTimeout
				return true
			}
		}
	}
	return false
}

type Guilds map[string]*Guild
type Players map[riotapi.Puuid]*Player

type Bot struct {
	token                       string
	database                    DatabaseBot
	mu                          *sync.RWMutex
	riotapi                     *riotapi.RiotApi
	guilds                      Guilds
	players                     Players
	riotapiHousekeepingExecutor common.TimedExecutor
	offlineThreshold            time.Duration
	offlineTimeout              time.Duration
	onlineTimeout               time.Duration
	mainCycle                   time.Duration
	discordgo                   *discordgo.Session
}

func NewBot(
	token string,
	dbFilename string,
	riotapiHousekeepingTimeout time.Duration,
	offlineThreshold time.Duration,
	offlineTimeout time.Duration,
	onlineTimeout time.Duration,
	mainCycle time.Duration,
	riotapi *riotapi.RiotApi) (Bot, error) {

	var bot Bot

	bot.token = token
	// Database
	bot.database = NewDatabaseBot(dbFilename)
	bot.mu = &sync.RWMutex{}
	// Initialise values from the database if present
	bot.guilds = bot.database.GetGuilds()
	bot.players = bot.database.GetPlayers()
	// Housekeeping for the riot API
	bot.riotapiHousekeepingExecutor = common.NewTimedExecutor(riotapiHousekeepingTimeout, bot.riotapiHousekeeping)
	// Timeouts for online and offline players
	bot.offlineThreshold = offlineThreshold
	bot.offlineTimeout = offlineTimeout
	bot.onlineTimeout = onlineTimeout
	// Set default (offline) timeouts for all players
	for _, player := range bot.players {
		player.StopWatch.Timeout = offlineTimeout
	}
	// Main loop cycle
	bot.mainCycle = mainCycle
	// Riot API
	bot.riotapi = riotapi

	return bot, nil

}

func connectWithRetry(dg *discordgo.Session) {

	// This is executed only when the application first starts, and attempts connection
	// to the discord gateway repeatedly until the connection is first established. After that,
	// supposedly the discordgo package will keep the connection alive and retry connections
	log.Info().Msg("Attempting connection to the Discord gateway")
	for {
		err := dg.Open()
		if err == nil {
			log.Info().Msg("Connected to Discord gateway successfully")
			return
		}

		log.Warn().Msg(fmt.Sprintf("Discord connection failed: %v. Retrying in 10s...", err))
		time.Sleep(10 * time.Second)
	}
}

func (bot *Bot) Run() error {
	// Create session
	var err error
	bot.discordgo, err = discordgo.New("Bot " + bot.token)
	if err != nil {
		return fmt.Errorf("could not create discord session")
	}

	// Event handler
	bot.discordgo.AddHandler(bot.Receive)

	// Open session
	connectWithRetry(bot.discordgo)
	defer bot.discordgo.Close()

	// keep bot running until there is an OS interruption
	log.Info().Msg("Running")
	go bot.loop()

	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
	<-c

	return nil
}

func (bot *Bot) Receive(discord *discordgo.Session, message *discordgo.MessageCreate) {

	// Reject my own messages
	if message.Author.ID == discord.State.User.ID {
		log.Debug().Msg("Rejecting my own message")
		return
	}

	// Ignore messages from private channels
	if message.GuildID == "" {
		log.Info().Msg("Ignoring private message")
		content := "For the time being, I am ignoring private messages"
		bot.sendResponsesToChannel([]Response{ResponseString{content}}, message.ChannelID)
		return
	}

	// Parse the input provided and call the appropriate function
	log.Debug().Msg(fmt.Sprintf("Received message: %s", message.Content))
	parseResult := Parse(message.Content)

	switch parseResult.parseid {
	case PARSEID_NO_BOT_PREFIX:
		log.Debug().Msg("Rejecting message not intended for the bot")
		return
	case PARSEID_OK:
		log.Info().Msg(fmt.Sprintf("Command understood: %s", message.Content))
		var responses []Response

		// Extract the name of the channel that has sent the message
		channelName, err := bot.getChannelName(discord, message.GuildID, message.ChannelID)
		if err != nil {
			log.Error().Msg(fmt.Sprintf("Could not extract channel name for channel id %s", message.ChannelID))
			return
		}

		// Register the guild if it's the first time I see it
		if bot.addGuildIfMissing(message.GuildID, message.ChannelID) {
			log.Info().Msg(fmt.Sprintf("Initialising guild %s", message.GuildID))
			// Send welcome message
			log.Info().Msg("Sending welcome message")
			responses = append(responses, Welcome(channelName)...)
		}

		// If this is an old guild, take the opportunity to check if the
		// configured channel still exists. If it does not, update the channel
		// and notify
		configuredChannelId, ok := bot.guildChannelId(message.GuildID)
		if !ok {
			log.Error().Msg(fmt.Sprintf("Could not find guild %s after initialisation", message.GuildID))
			return
		}
		configuredChannelName, err := bot.getChannelName(discord, message.GuildID, configuredChannelId)
		if err != nil {
			bot.setGuildChannel(message.GuildID, message.ChannelID)
			configuredChannelName = channelName
			log.Info().Msg(fmt.Sprintf("Updated the channel for in-game messages to %s", channelName))
			responses = append(responses, ChannelDisappeared(channelName)...)
		}

		// Decide according to the command
		switch parseResult.command {
		case COMMAND_REGISTER:
			switch riotid := parseResult.arguments.(type) {
			default:
				panic(fmt.Sprintf("unexpected type of riot id %T", riotid))
			case riotapi.RiotId:
				responses = append(responses, bot.register(riotid, message.GuildID)...)
			}
		case COMMAND_UNREGISTER:
			switch riotid := parseResult.arguments.(type) {
			default:
				panic(fmt.Sprintf("unexpected type of riot id %T", riotid))
			case riotapi.RiotId:
				responses = append(responses, bot.unregister(riotid, message.GuildID)...)
			}
		case COMMAND_RANK:
			switch riotid := parseResult.arguments.(type) {
			default:
				panic(fmt.Sprintf("unexpected type of riot id %T", riotid))
			case riotapi.RiotId:
				responses = append(responses, bot.rank(riotid)...)
			}
		case COMMAND_CHANNEL:
			switch channelName := parseResult.arguments.(type) {
			default:
				panic(fmt.Sprintf("unexpected type of channel name %T", channelName))
			case string:
				responses = append(responses, bot.channel(discord, channelName, message.GuildID)...)
			}
		case COMMAND_STATUS:
			responses = append(responses, bot.status(discord, message.GuildID, configuredChannelName)...)
		case COMMAND_HELP:
			responses = append(responses, HelpMessage()...)
		default:
			panic(fmt.Sprintf("Command %d is not one of the possible ones", parseResult.command))
		}
		bot.sendResponsesToChannel(responses, message.ChannelID)
	default:

		// The command is invalid input, so it contains an error message
		errorMessage := parseResult.errorMessage
		log.Info().Msg(fmt.Sprintf("Wrong input: '%s'. Reason: %s", message.Content, errorMessage))
		bot.sendResponsesToChannel(InputNotValid(errorMessage), message.ChannelID)
	}
}

func (bot *Bot) sendResponsesToChannel(responses []Response, channelId string) {
	for _, response := range responses {
		response.Send(channelId, bot.discordgo)
	}
}

// When sending responses to a guild, find the channel id configured
// for the guild, and send the responses to that channel
func (bot *Bot) sendResponsesToGuild(responses []Response, guildid string) {

	// Get the channel id for this guild
	channelid, ok := bot.guildChannelId(guildid)
	if !ok {
		log.Warn().Msg(fmt.Sprintf("Could not send response because guild %s is not registered", guildid))
		return
	}
	bot.sendResponsesToChannel(responses, channelid)
}

func (bot *Bot) register(riotid riotapi.RiotId, guildid string) []Response {

	// Get the puuid
	puuid, err := bot.riotapi.GetPuuid(riotid)
	if err != nil {
		log.Info().Err(err)
		return NoResponseRiotApi(riotid)
	}

	// Check if player already belongs to the guild
	if bot.isPlayerRegisteredInGuild(puuid, guildid) {
		log.Info().Msg(fmt.Sprintf("Player %s is already registered in guild %s", riotid, guildid))
		return PlayerAlreadyRegistered(riotid)
	}

	// Get rank of this player
	leagues, err := bot.riotapi.GetLeagues(puuid)
	if err != nil {
		log.Info().Err(err)
		return NoResponseRiotApi(riotid)
	}

	// Now it's safe to register the player
	log.Info().Msg(fmt.Sprintf("Registering player %s in guild %s", &riotid, guildid))
	if !bot.addPlayerToGuild(puuid, guildid) {
		log.Info().Msg(fmt.Sprintf("Player %s is already registered in guild %s", riotid, guildid))
		return PlayerAlreadyRegistered(riotid)
	}

	// Send the final message
	log.Info().Msg(fmt.Sprintf("Player %s has been registered in guild %s", &riotid, guildid))
	return []Response{
		PlayerRegistered(riotid),
		PlayerRank(riotid, leagues),
	}
}

func (bot *Bot) unregister(riotid riotapi.RiotId, guildid string) []Response {

	// Get the puuid
	puuid, err := bot.riotapi.GetPuuid(riotid)
	if err != nil {
		log.Info().Err(err)
		return NoResponseRiotApi(riotid)
	}

	// Check if the player was registered in the guild
	if !bot.isPlayerRegisteredInGuild(puuid, guildid) {
		log.Info().Msg(fmt.Sprintf("Player %s was not previously registered in guild %s", riotid, guildid))
		return PlayerNotPreviouslyRegistered(riotid)
	}

	// Check if the player was registered in the guild
	log.Info().Msg(fmt.Sprintf("Unregistering player %s from guild id %s", riotid, guildid))
	removedFromGuild, removedCompletely := bot.removePlayerFromGuild(puuid, guildid)
	if !removedFromGuild {
		log.Info().Msg(fmt.Sprintf("Player %s was not previously registered in guild %s", riotid, guildid))
		return PlayerNotPreviouslyRegistered(riotid)
	}
	log.Info().Msg(fmt.Sprintf("Player %s has been unregistered from guild %s", riotid, guildid))
	// If the player does not belong to ANY guild, remove it completely
	if removedCompletely {
		log.Info().Msg(fmt.Sprintf("Player %s removed completely", riotid))
	}
	return PlayerUnregistered(riotid)
}

func (bot *Bot) rank(riotid riotapi.RiotId) []Response {

	// Get the puuid
	puuid, err := bot.riotapi.GetPuuid(riotid)
	if err != nil {
		log.Info().Err(err)
		return NoResponseRiotApi(riotid)
	}

	// Get the rank of this player
	leagues, err := bot.riotapi.GetLeagues(puuid)
	if err != nil {
		log.Info().Err(err)
		return NoResponseRiotApi(riotid)
	}

	// Send the final message
	log.Info().Msg(fmt.Sprintf("Sending rank of player %s", riotid))
	return []Response{PlayerRank(riotid, leagues)}
}

func (bot *Bot) channel(discord *discordgo.Session, channelName string, guildid string) []Response {

	// Try to find the id from the channel name
	channelId, err := bot.getChannelId(discord, guildid, channelName)
	if err != nil {
		log.Info().Msg(fmt.Sprintf("Could not extract channel id from channel name %s", channelName))
		return ChannelDoesNotExist(channelName)
	}

	// We have a new channel to send messages to
	log.Info().Msg(fmt.Sprintf("Changing channel used by guild %s to %s", guildid, channelName))
	bot.setGuildChannel(guildid, channelId)
	return ChannelChanged(channelName)
}

func (bot *Bot) status(discord *discordgo.Session, guildid string, configuredChannelName string) []Response {

	// Create list of player names in this guild
	riotIds := []riotapi.RiotId{}
	for _, puuid := range bot.guildPuuids(guildid) {
		riotid, err := bot.riotapi.GetRiotId(puuid)
		if err != nil {
			panic(err)
		}
		riotIds = append(riotIds, riotid)
	}
	return StatusMessage(riotIds, configuredChannelName)
}

func (bot *Bot) loop() {

	for {
		// TODO: investigate if this wait time needs to be updated according to
		// the number of users registered
		var wg sync.WaitGroup
		wg.Add(1)
		go func() {
			defer wg.Done()
			time.Sleep(bot.mainCycle)
		}()
		wg.Wait()
		// Housekeeping of the Riot API if necessary
		bot.riotapiHousekeepingExecutor.Execute()
		// Find a player to check during this iteration
		puuid, found := bot.selectPlayerToCheck()
		if !found {
			continue
		}
		riotid, err := bot.riotapi.GetRiotId(puuid)
		if err != nil {
			log.Error().Err(err)
			continue
		}
		// We have a player to check
		log.Debug().Msg(fmt.Sprintf("Checking in game status of player %s", riotid))
		spectator, err := bot.riotapi.GetSpectator(puuid)
		// At this point, a request has been performed in some form,
		// so the stopwatch needs to be reset
		if !bot.startPlayerStopwatch(puuid) {
			continue
		}
		if err != nil {
			if errors.Is(err, common.ErrNotFound) {
				log.Debug().Msg(fmt.Sprintf("Spectator data for player %s is not available", riotid))
				if bot.updatePlayerCheckTimeout(puuid, false) {
					log.Info().Msg(fmt.Sprintf("Player %s is now offline", riotid))
				}
			} else {
				log.Warn().Err(err).Msg(fmt.Sprintf("Could not check spectator data for player %s", riotid))
			}
			continue
		}
		// Player is online, update timeout if needed
		if bot.updatePlayerCheckTimeout(puuid, true) {
			log.Info().Msg(fmt.Sprintf("Player %s is now online", &riotid))
		}
		// Get the guilds where this player is registered
		for _, target := range bot.guildNotificationTargets(puuid, spectator.GameId) {
			log.Info().Msg(fmt.Sprintf("Sending spectator message for player %s and game %d in guild %s", &riotid, spectator.GameId, target.guildid))
			// Build the complete response
			responses := InGameMessage(puuid, riotid, spectator)
			bot.sendResponsesToChannel(responses, target.channelid)
			// Update last informed game id
			bot.setLastInformedGameId(puuid, target.guildid, spectator.GameId)
		}
	}
}

func (bot *Bot) selectPlayerToCheck() (riotapi.Puuid, bool) {

	log.Debug().Msg("Selecting player to check in this iteration")
	var longestTimeStopped time.Duration
	var puuid riotapi.Puuid
	for _, player := range bot.playerCheckStates() {
		riotid, err := bot.riotapi.GetRiotId(player.puuid)
		if err != nil {
			panic(err)
		}
		// Time the stopwatch has been stopped for this player
		if player.stopped {
			if player.elapsed > longestTimeStopped {
				// This player is available to be checked,
				// and has been stopped more than the others,
				// so it's our best candidate
				log.Debug().Msg(fmt.Sprintf("- %s: player has been stopped for longer", &riotid))
				longestTimeStopped = player.elapsed
				puuid = player.puuid
			} else {
				// This player is available to check, but has been checked
				// more recently than others
				log.Debug().Msg(fmt.Sprintf("- %s: player has been checked more recently", &riotid))
			}
		} else {
			log.Debug().Msg(fmt.Sprintf("- %s: timeout still not reached", &riotid))
		}
	}
	// We have selected the best option to check
	if puuid == "" {
		log.Debug().Msg("Not checking any player during this iteration")
		return puuid, false
	} else {
		log.Debug().Msg(fmt.Sprintf("Checking player with puuid %s in this iteration", puuid))
		return puuid, true
	}
}

// Get channel name, provided a channel id, from the discord api
func (bot *Bot) getChannelName(discord *discordgo.Session, guildid string, channelid string) (string, error) {

	channels, err := discord.GuildChannels(guildid)
	if err != nil {
		return "", fmt.Errorf("could not extract list of channels of guild id %s", guildid)
	}
	for _, ch := range channels {
		if ch.ID == channelid {
			return ch.Name, nil
		}
	}
	return "", fmt.Errorf("no channel name found for channel id %s", channelid)
}

func (bot *Bot) getChannelId(discord *discordgo.Session, guildid string, channelName string) (string, error) {

	channels, err := discord.GuildChannels(guildid)
	if err != nil {
		return "", fmt.Errorf("could not extract list of channels of guild id %s", guildid)
	}
	for _, ch := range channels {
		if ch.Name == channelName {
			return ch.ID, nil
		}
	}
	return "", fmt.Errorf("no channel id found for channel name %s", channelName)
}

func (bot *Bot) riotapiHousekeeping() {
	log.Info().Msg("Performing Riot API housekeeping")
	bot.riotapi.Housekeeping(bot.registeredPuuids())
}
